### PR TITLE
[Models] Add DSpark speculator (DFlash + Markov + confidence head)

### PR DESCRIPTION
**Pull request by mgoin**
## Summary
Adds **DSpark** (based on the new method in [DeepSpec](https://github.com/deepseek-ai/DeepSpec)), which extends DFlash with a low-rank Markov logit-bias head (intra-block token dependency) and a per-position confidence head, trained with a TV-dominant loss. DFlash with the heads disabled (`--markov-rank 0`) reproduces plain DFlash.

- `src/speculators/models/dspark/` — config, model, Markov/confidence heads, loss
- `scripts/train.py` — `--speculator-type dspark` + markov/confidence/loss-weight flags
- unit tests + an online training example

## Results
Qwen3-8B, matched 20k-ShareGPT online runs (3 epochs, seq 2048), validation per-position acceptance:

| run | full_acc | pos1 | pos4 | pos7 |
|---|---|---|---|---|
| DFlash (kl_div) | 0.309 | 0.665 | 0.243 | 0.115 |
| + TV loss (markov off) | 0.341 | 0.690 | 0.282 | 0.136 |
| + Markov head (DSpark) | 0.415 | 0.689 | 0.369 | 0.255 |

The Markov head contributes most of the gain and grows with position (the suffix-decay fix); the TV loss adds a smaller, roughly uniform lift.

## Testing
`pytest tests/unit/models/test_dspark_*` → 13 passed; `ruff` clean. Trained DSpark/DFlash drafts for Qwen3-0.6B/4B/8B online via the example script.

AI assistance (Claude) was used; the submitter reviewed every line. No existing PR covers DSpark.

### CODE REVIEW

File: src/speculators/models/dspark/core.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        """DSpark owns its multi-term loss; only the decay gamma flows through.
+
+        The ``ce``/``tv``/``confidence`` weights are read from the saved config, and
+        ``--loss-fn`` is intentionally ignored for DSpark.
+        """
+        gamma = kwargs.get("dflash_decay_gamma", 4.0)
+        return {"gamma": gamma}, {"gamma": gamma}
+
+    @conditional_torch_compile
+    def forward(
```
Comment (shanjiaz) [inline src/speculators/models/dspark/core.py:94]: This is almost identical to DFlash's forward pass, can we maybe refactor DFlash's forward pass to extract a `_backbone_forward` method (or any other reasonable name) on DFlash that returns the intermediate hidden states, logits, targets, aligned loss mask, block indices), so DSpark's forward only contains the Markov bias, confidence head, and its own loss call. Similarly for `from_training_args`, the shared config kwargs could be extracted into a helper.

File: src/speculators/models/dspark/core.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        anchored_block_indices = get_base_indices_for_anchored_blocks(
+            anchor_positions, self.block_size
+        )  # [num_anchors*block_size]
+
+        with torch.no_grad():
+            verifier_logits = self.verifier_lm_head(
+                self.verifier_norm(verifier_last_hidden_states)
+            )
+            verifier_logits = torch.roll(verifier_logits, 1, dims=1)
+            targets = verifier_logits[:, anchored_block_indices]
```
[the remark this answers is not in the document]
Comment (shanjiaz) [reply src/speculators/models/dspark/core.py]: You're wrong, in model.py we already reduce vocab size. Both lm_head and verifier_lm_head output draft_vocab_size dimensions, not verifier vocab size. And in load_verifier_weights (model.py:161-169), when use_draft_vocab is true, the verifier's full lm_head weight is sliced down to only the draft vocab rows via the t2d mask before being loaded into both heads.

So by the time forward runs, verifier_lm_head already projects into draft-vocab space. The targets at dflash/core.py:328-334 are in draft-vocab space, same as the draft logits. No mismatch — the mapping is handled at weight-loading time, not at forward time.

**Review by shanjiaz — COMMENTED** [review]
Looks great! Thanks for running the experiments. Just one small comment about reusing DFlash's functions for maintainability.

**Comment by WindChimeRan** [comment]
### DSpark (Qwen3-8B): training-data policy vs. validation set

Our speculator train & eval is default on **off-policy** (not response re-generation). This is not ideal

| Training responses              | Validation set | full_acc | pos1  | pos4  | pos7  |
|---------------------------------|----------------|----------|-------|-------|-------|
| off-policy (ShareGPT) — reported| off-policy     | 0.415    | 0.689 | 0.369 | 0.255 |
| off-policy (ShareGPT)           | on-policy      | 0.280    | 0.550 | 0.228 | 0.156 |
| on-policy, single-turn          | on-policy      | 0.374    | 0.609 | 0.334 | 0.250 |
| on-policy, multi-turn   #632        | on-policy      | 0.405    | 0.648 | 0.365 | 0.272 |

The bottom three rows share the same held-out on-policy validation set, so they are directly comparable.

- On the on-policy (deployment-like) validation set, on-policy training beats   off-policy at every position: **0.405 vs 0.280 full_acc (+45%)**.
- The off-policy recipe scores 0.415 on its own off-policy validation split but   0.280 when validated on on-policy generations
- Most of the gain comes from using on-policy responses; multi-turn regeneration  adds a further improvement over single-turn.

**Comment by mgoin** [comment]
^ I did full training runs with response regen https://huggingface.co/mgoin/Qwen3-8B-speculator.dspark-reasoning so fully agree with the impact of on-policy :)

**Comment by WindChimeRan** [comment]
@mgoin then I think your reported setting was single-turn on policy train, single-turn val, if you were using the main branch train, not cherrypicking the #632? 

---

I'll prioritize making on-policy train as default, and try to land some quick win first.

**Review by shanjiaz — APPROVED** [review]
Discussed offline. Some small concerns around lossConfig, otherwise looks great!

**Comment by SuQinghang** [comment]
Hello, I have trained a dspark drafter for my own model and deployed it with vllm0.25.0. However, the accept length is extremely low, about 1 only.  
Besides, I also test the performance of Qwen3-8B + DSpark (https://huggingface.co/mgoin/Qwen3-8B-speculator.dspark) on GSM8K, the accept length is 2 tokens.
I want to know about the commands to deploy dspark ckpts trained by this pr with vllm. Following is my dspark config.json and vllm serve commands

dspark config.json for my model

```json 
{
  "architectures": [
    "DSparkDraftModel"
  ],
  "auto_map": {
    "": "config.DSparkSpeculatorConfig"
  },
  "aux_hidden_state_layer_ids": [
    2,
    31,
    59
  ],
  "block_size": 8,
  "confidence_head_with_markov": true,
  "draft_vocab_size": 32000,
  "dtype": "bfloat16",
  "enable_confidence_head": true,
  "markov_head_type": "vanilla",
  "markov_rank": 256,
  "mask_token_id": 167588,
  "sample_from_anchor": true,
  "sliding_window_non_causal": false,
  "speculators_config": {
    "algorithm": "dspark",
    "default_proposal_method": "greedy",
    "proposal_methods": [
      {
        "accept_tolerance": 0.0,
        "proposal_type": "greedy",
        "speculative_tokens": 8,
        "verifier_accept_k": 1
      }
    ],
    "verifier": {
      "architectures": [
        "LlamaForCausalLM"
      ],
      "name_or_path": "/path/to/mymodel"
    }
  },
  "speculators_model_type": "dspark",
  "speculators_version": "0.5.0",
  "target_hidden_size": null,
  "tie_word_embeddings": false,
  "transformer_layer_config": {
    "attention_bias": false,
    "attention_dropout": 0.0,
    "bos_token_id": null,
    "eos_token_id": null,
    "head_dim": 128,
    "hidden_act": "silu",
    "hidden_size": 6144,
    "initializer_range": 0.008944,
    "intermediate_size": 24576,
    "layer_types": [
      "sliding_attention",
      "sliding_attention",
      "sliding_attention"
    ],
    "max_position_embeddings": 262144,
    "max_window_layers": 28,
    "model_type": "qwen3",
    "num_attention_heads": 48,
    "num_hidden_layers": 3,
    "num_key_value_heads": 8,
    "pad_token_id": null,
    "rms_norm_eps": 1e-06,
    "rope_parameters": {
      "rope_theta": 5000000.0,
      "rope_type": "default"
    },
    "sliding_window": 2048,
    "tie_word_embeddings": false,
    "use_cache": true,
    "use_sliding_window": true,
    "vocab_size": 167936
  },
  "transformers_version": "5.6.2"
}
```
vllm serve command for my model
```bash
MODEL=/path/to/mymodel
DSPARK_PATH=/path/to/dspark/ckpt

CUDA_VISIBLE_DEVICES=0,1 \
vllm serve "$MODEL" \
  --port 8001 \
  --max-model-len 64k \
  --tensor-parallel-size 2 \
  --gpu-memory-utilization 0.7 \
  --async-scheduling \
  --additional-config '{"enable_cpu_binding":true}' \
  --speculative-config '{
    "model": "$DSPARK_PATH",
    "method": "dspark",
    "num_speculative_tokens": 7,
    "draft_sample_method": "greedy"
  }'
```

vllm serve command for Qwen3-8B
```bash
MODEL=/path/to/Qwen3-8B

CUDA_VISIBLE_DEVICES=0,1 \
vllm serve "$MODEL" \
  --port 8001 \
  --tensor-parallel-size 2 \
  --gpu-memory-utilization 0.7 \
  --async-scheduling \
  --additional-config '{"enable_cpu_binding":true}' \
  --speculative-config '{
    "model": "/path/to/qwen3-dspark",
    "method": "dspark",
    "num_speculative_tokens": 7,
    "draft_sample_method": "greedy"
  }'
  ```

**Comment by orestis-z** [comment]
@SuQinghang Possibly related: I hit this same issue in my ablation runs (#757). It was caused by `dspark_bonus_anchor = True` being hardcoded in vLLM's `algos.py` ([vllm-project/vllm#47093](https://github.com/vllm-project/vllm/pull/47093)), which forced `sample_from_anchor=False` regardless of the checkpoint config. This has since been fixed on vLLM main — `dspark_bonus_anchor` was replaced with a proper `sample_from_anchor` passthrough.

**Comment by HaizhouPeng** [comment]
@SuQinghang I also meeted the same problem. Did you resolve it through the reasons mentioned by @orestis-z  ?

**Comment by SuQinghang** [comment]
@HaizhouPeng  Hi, I have solved this problem by training with the latest main branch which merged the pr #798 #805 #806.

**Comment by HaizhouPeng** [comment]
@SuQinghang Thanks, i will try then
